src/lib/FreeC/Backend/Coq/Converter/TypeDecl.hs

```diff
@@ -218,6 +224,71 @@ convertDataDecl (IR.DataDecl _ (IR.DeclIdent _ name) typeVarDecls conDecls) =
         rhs
       )
 
+  -- | Generates an induction scheme for the data type.
+  generateInductionScheme :: Converter Coq.Sentence
+  generateInductionScheme = localEnv $ do
+    Just tIdent <- inEnv $ lookupIdent IR.TypeScope name
+    -- Create variables and binders.
+    let generateArg argType = do
+          ident <- freshCoqQualid freshArgPrefix
+          return $ (ident, Coq.typedBinder Coq.Explicit [ident] argType)
+    (tvarIdents, tvarBinders) <- convertTypeVarDecls' Coq.Explicit typeVarDecls
+    (propIdent , propBinder ) <- generateArg
+      (Coq.Arrow (genericApply tIdent [] [] (map Coq.Qualid tvarIdents))
+                 (Coq.Sort Coq.Prop)
+      )
+    (_hIdents, hBinders) <- mapAndUnzipM (generateInductionCase propIdent)
+                                         conDecls
+    (valIdent, valBinder) <- generateArg
+      (genericApply tIdent [] [] (map Coq.Qualid tvarIdents))
+    -- Stick everything together.
+    let schemeName = Coq.bare $ fromJust (Coq.unpackQualid tIdent) ++ "_Ind"
```

Comment:
Manually generating a new Coq identifier is not safe. Use `freshCoqQualid` instead.

Reply:
The function `freshCoqQualid` seems to be appending always an underscore and a value to that name. As the `schemeName` is used by the user in the proofs it would be nice if the name was untouched unless it is neccessary. Would `renameQualid` be an option here?

Reply:
Unfortunatly, `renameQualid` does not suffice since it doesn't insert an environment entry. However, `renameQualid` is used by `freshCoqQualid` internally. `freshCoqQualid` generates a Haskell name of the form `Prefix@Index`, uses `renameQualid` to convert it to a Coq identifier of the form `Prefix_Index` and finally inserts an environment entry such that the Coq identifier is not used again. The `@Index` postfix is only needed to ensure that the Haskell identifier (that identifies the environment entry) does not collide with a user defined identifier. You could try to strip the `@Index` suffix from identifiers passed to `renameQualid`. [`freshHaskellIdent`](https://github.com/FreeProving/free-compiler/blob/8779013877645b9547bb36c93fa1df9856b2b1aa/src/lib/FreeC/Environment/Fresh.hs#L86) does something similar. I think that this should be safe but you may have to update a lot of test cases.

Reply:
I changed `renameIdent` the way you suggested it for `renameQualid` because `renameQualid` calls this function and they should produce the same names.